src/video/wayland/SDL_waylandwindow.c

```diff
@@ -1502,6 +1503,77 @@ static const struct wp_fractional_scale_v1_listener fractional_scale_listener =
     handle_preferred_fractional_scale
 };
 
+static struct xdg_toplevel *GetToplevelForWindow(SDL_WindowData *wind)
+{
+    if (wind) {
+        /* Libdecor crashes on attempts to unset the parent by passing null, which is allowed by the
+         * toplevel spec, so just use the raw xdg-toplevel instead (that's what libdecor does
+         * internally anyways).
+         */
+#ifdef HAVE_LIBDECOR_H
+        if (wind->shell_surface_type == WAYLAND_SURFACE_LIBDECOR && wind->shell_surface.libdecor.frame) {
+            return libdecor_frame_get_xdg_toplevel(wind->shell_surface.libdecor.frame);
+        } else
+#endif
+            if (wind->shell_surface_type == WAYLAND_SURFACE_XDG_TOPLEVEL && wind->shell_surface.xdg.roleobj.toplevel) {
+                return wind->shell_surface.xdg.roleobj.toplevel;
+            }
+    }
+
+    return NULL;
+}
+
+static void handle_xdg_session_created(void *data, struct xdg_session_v1 *xdg_session_v1, const char *id)
+{
+    SDL_SetStringProperty(SDL_GetGlobalProperties(), SDL_PROP_GLOBAL_VIDEO_WAYLAND_SESSION_ID_STRING, id);
+}
+
+static void handle_xdg_session_restored(void *data, struct xdg_session_v1 *xdg_session_v1)
+{
+    // NOP
+}
+
+static void handle_xdg_session_replaced(void *data, struct xdg_session_v1 *xdg_session_v1)
+{
+    // NOP
+}
+
+static const struct xdg_session_v1_listener xdg_session_listener = {
+    .created  = handle_xdg_session_created,
+    .restored = handle_xdg_session_restored,
+    .replaced = handle_xdg_session_replaced
+};
+
+static void Wayland_CreateSession(SDL_VideoData *data)
+{
+    if (!data->xdg_session_manager_v1) {
+        return;
+    }
+
+    // Register a new session, if one does not yet exist.
+    if (!data->xdg_session_v1) {
+        const char *session_id = SDL_GetStringProperty(SDL_GetGlobalProperties(), SDL_PROP_GLOBAL_VIDEO_WAYLAND_SESSION_ID_STRING, NULL);
+        if (session_id && *session_id == '\0') {
+            session_id = NULL;
+        }
+        data->xdg_session_v1 = xdg_session_manager_v1_get_session(data->xdg_session_manager_v1, XDG_SESSION_MANAGER_V1_REASON_LAUNCH, session_id);
+
+        xdg_session_v1_add_listener(data->xdg_session_v1, &xdg_session_listener, data);
+
+        // Need to round trip as the session ID is needed now.
+        WAYLAND_wl_display_roundtrip(data->display);
```

Comment:
FYI the immediate ID response guarantee was dropped from the protocol spec, so you can end up with null id after this roundtrip.

ref: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/18#note_2353465